### PR TITLE
TcpSocket specialized split

### LINKED ISSUE #174: Consider a specialized split fn for TcpStream
The type can safely support a read and a write task without coordination.

### LINKED ISSUE #852: WriteHalf<TcpStream>::shutdown won't shutdown the TcpStream
## Version

tokio 0.1.14

## Platform

Linux archlinux 4.19.13-1-lts #1 SMP Sun Dec 30 07:38:47 CET 2018 x86_64 GNU/Linux

## Subcrates

tokio-tcp

## Description

I'm trying to implement a TCP proxy and I know there is an example. I noticed that calling the shutdown method of `WriteHalf<TcpStream>` does nothing. I find this after some research:

```rust
impl<'a> AsyncWrite for &'a TcpStream {
    fn shutdown(&mut self) -> Poll<(), io::Error> {
        Ok(().into())
    }
```

My question is that why this shutdown does nothing instead of calling TcpStream::shutdown(self, Shutdown::Write) since it allow us write something like this:

```rust

                    let (client_reader, client_writer) = client.split();
                    let (server_reader, server_writer) = server.split();

                    let client_to_server = copy(client_reader, server_writer)
                        .and_then(|(n, _, mut server_writer)| {
                            println!("copied {} bytes to server", n);
                            server_writer.shutdown()
                        });
                    let server_to_client = copy(server_reader, client_writer)
                        .and_then(|(n, _, mut client_writer)| {
                            println!("copied {} bytes to client", n);
                            client_writer.shutdown()
                        });
                    client_to_server.join(server_to_client)

```

### PR DESCRIPTION
**Pull request by sopium**
A specialized split is implemented for `TcpStream`(close #174, related: #1108).

Compared to the generic `AsyncRead::split`, the specialized split gives handles (halves) that are faster and smaller, because they do not use locks.

The write half handle properly implements the shutdown operation, by shutting down the underlying stream in the "write" direction (close #852).

The handles also implement `AsRef<TcpStream>`, allowing access to methods of the `TcpStream` that takes `&self`: to get local and peer addresses, to get/set various socket options, to shutdown the socket, etc. (Close #852 even if write half shutdown does not call stream shutdown.)

## Future Work

Other IO resources should have similar specialized split support as well.

### CODE REVIEW

**Comment by sopium** [comment]
I'd like to have some feedback. I will finish the docs and `reunite`, `ReuniteError` if you agree that this is a valid design.

**Comment by jonhoo** [comment]
This seems like a great idea, and I think the approach looks reasonable. There are some docs missing, but you're aware of that :) I've left a couple of smaller inline comments.

File: tokio-tcp/Cargo.toml [hunk 1]
```diff
@@ -23,6 +23,7 @@ publish = false
 
 [features]
 incoming = ["futures-core-preview"]
+default = ["incoming"]
```
Comment (jonhoo) [inline tokio-tcp/Cargo.toml]: This seems like an unrelated change?

File: tokio/Cargo.toml [hunk 2]
```diff
@@ -98,3 +98,4 @@ num_cpus = "1.0"
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 time = "0.1"
+futures-preview = "0.3.0-alpha.16"
```
Comment (jonhoo) [inline tokio/Cargo.toml]: This seems like a larger change that shouldn't be made in the context of this PR

File: tokio/examples/proxy.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+//!     cargo run --example echo
+//!
+//! And finally this in another terminal
+//!
+//!     cargo run --example connect 127.0.0.1:8081
+//!
+//! This final terminal will connect to our proxy, which will in turn connect to
+//! the echo server, and you'll be able to see data flowing between them.
+
+#![feature(async_await)]
```
Comment (jonhoo) [inline tokio/examples/proxy.rs]: I think this should probably be submitted as a separate PR that move the proxy example to use async/await syntax (see also https://github.com/tokio-rs/tokio/issues/1201).

**Comment by sopium** [comment]
OK, I will not include the proxy example in this PR. Post it here:

```rust
//! A proxy that forwards data to another server and forwards that server's
//! responses back to clients.
//!
//! Because the Tokio runtime uses a thread pool, each TCP connection is
//! processed concurrently with all other TCP connections across multiple
//! threads.
//!
//! You can showcase this by running this in one terminal:
//!
//!     cargo run --example proxy
//!
//! This in another terminal
//!
//!     cargo run --example echo
//!
//! And finally this in another terminal
//!
//!     cargo run --example connect 127.0.0.1:8081
//!
//! This final terminal will connect to our proxy, which will in turn connect to
//! the echo server, and you'll be able to see data flowing between them.

#![feature(async_await)]

use futures::future::try_join;
use futures::prelude::StreamExt;
use std::env;
use std::net::SocketAddr;
use tokio;
use tokio::io::AsyncReadExt;
use tokio::net::tcp::split::{TcpStreamReadHalf, TcpStreamWriteHalf};
use tokio::net::tcp::{TcpListener, TcpStream};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let listen_addr = env::args().nth(1).unwrap_or("127.0.0.1:8081".to_string());
    let listen_addr = listen_addr.parse::<SocketAddr>()?;

    let server_addr = env::args().nth(2).unwrap_or("127.0.0.1:8080".to_string());
    let server_addr = server_addr.parse::<SocketAddr>()?;

    // Create a TCP listener which will listen for incoming connections.
    let socket = TcpListener::bind(&listen_addr)?;
    println!("Listening on: {}", listen_addr);
    println!("Proxying to: {}", server_addr);

    let mut incoming = socket.incoming();
    loop {
        let stream = incoming.next().await.unwrap()?;
        tokio::spawn(async move {
            match proxy_client(stream, server_addr).await {
                Err(e) => {
                    eprintln!("Error: {}", e);
                }
                _ => (),
            }
        });
    }
}

async fn proxy_client(
    client_stream: TcpStream,
    server_addr: SocketAddr,
) -> Result<(), std::io::Error> {
    let server_stream = TcpStream::connect(&server_addr).await?;

    // Create separate read/write handles for the TCP clients that we're
    // proxying data between.
    //
    // Note that while you can use `AsyncRead::split` for this operation,
    // `TcpStream::split` gives you handles that are faster, smaller and allow
    // proper shutdown operations.
    let (client_r, client_w) = client_stream.split();
    let (server_r, server_w) = server_stream.split();

    let client_to_server = copy_shutdown(client_r, server_w);
    let server_to_client = copy_shutdown(server_r, client_w);

    // Run the two futures in parallel.
    let (l1, l2) = try_join(client_to_server, server_to_client).await?;
    println!("client wrote {} bytes and received {} bytes", l1, l2);
    Ok(())
}

// Copy data from a read half to a write half. After the copy is done we
// indicate to the remote side that we've finished by shutting down the
// connection.
async fn copy_shutdown(
    mut r: TcpStreamReadHalf,
    mut w: TcpStreamWriteHalf,
) -> Result<u64, std::io::Error> {
    let l = r.copy(&mut w).await?;

    // Use this instead after `shutdown` is implemented in `AsyncWriteExt`:
    // w.shutdown().await?;
    w.as_ref().shutdown(std::net::Shutdown::Write)?;

    Ok(l)
}
```

**Comment by liranringel** [comment]
This PR is really useful, thanks!
What do you think about something like `split3`, that will return a read half, a write half and an `ImmutableTcpStream` that will implement `AsRef<TcpStream>` so the async io requests could be canceled easily with `shutdown` from another task?

**Comment by sopium** [comment]
> This PR is really useful, thanks!
> What do you think about something like `split3`, that will return a read half, a write half and an `ImmutableTcpStream` that will implement `AsRef<TcpStream>` so the async io requests could be canceled easily with `shutdown` from another task?

What's your use case?

I think in rust async/await model individual IO resources do not need to have any special support for cancellation. If you no longer poll the future, it is "cancelled".

That said, with the current design, it is possible to shutdown the read half from the write half (or vice versa).

File: tokio-tcp/src/split.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+impl Error for ReuniteError {}
+
+impl TcpStreamReadHalf {
+    /// Attempts to put the two "halves" of a `TcpStream` back together and
+    /// recover the original stream. Succeeds only if the two "halves"
+    /// originated from the same call to `TcpStream::split`.
+    pub fn reunite(self, other: TcpStreamWriteHalf) -> Result<TcpStream, ReuniteError> {
+        if Arc::ptr_eq(&self.0, &other.0) {
+            drop(other);
+            Ok(Arc::try_unwrap(self.0).unwrap())
```
Comment (jonhoo) [inline tokio-tcp/src/split.rs]: Could you make this into an `expect` with an explanation for why it should always be ok?

File: tokio-tcp/Cargo.toml [hunk 5]
```diff
@@ -37,4 +37,4 @@ futures-core-preview = { version = "0.3.0-alpha.16", optional = true }
 [dev-dependencies]
 #env_logger = { version = "0.5", default-features = false }
 #net2 = "*"
-#tokio = { version = "0.2.0", path = "../tokio" }
+tokio = { version = "0.2.0", path = "../tokio" }
```
Comment (jonhoo) [inline tokio-tcp/Cargo.toml]: I don't _think_ you need this change any more?

Comment (sopium) [reply tokio-tcp/Cargo.toml]: It is needed for the test.

Comment (jonhoo) [reply tokio-tcp/Cargo.toml]: Ohh, I missed the `#[tokio::test]` annotation :+1:

File: tokio-tcp/src/split.rs (hunk 4 above)
Comment (jonhoo) [reply tokio-tcp/src/split.rs]: Ah, sorry, what I meant here was to explain why you are expecting this unwrap to never fail. I _believe_ it is because you only ever create two instances of the `Arc`, one for the reader and one for the writer, and those `Arc`s are never exposed externally. And so when you drop one here, the other one _must_ be the only remaining one.

**Comment by jonhoo** [comment]
Ah, sorry, what I meant about the `expect` was to explain why you are expecting this unwrap to never fail. I _believe_ it is because you only ever create two instances of the `Arc`, one for the reader and one for the writer, and those `Arc`s are never exposed externally. And so when you drop one here, the other one _must_ be the only remaining one.

**Comment by sopium** [comment]
> Ah, sorry, what I meant about the `expect` was to explain why you are expecting this unwrap to never fail. I _believe_ it is because you only ever create two instances of the `Arc`, one for the reader and one for the writer, and those `Arc`s are never exposed externally. And so when you drop one here, the other one _must_ be the only remaining one.

Yes, that's correct. I think it's a common pattern in `BiLock` and other split implementations.

**Comment by jonhoo** [comment]
@sopium Yeah, I think you're right, I'd just like to see that argument made in the code to make it clear why we think unwrapping/expecting there is okay :) Other than that I'm :+1: on this PR!

**Comment by sopium** [comment]
Sure, I added a comment roughly based on your comment. You explained it very well.

**Comment by carllerche** [comment]
Thanks :+1: I have some minor bikeshed level stuff, but nothing worth blocking merging 👍 I added notes to #1209 to track polish TODOs.
